pandas/tests/tslibs/test_conversion.py

```diff
@@ -66,3 +69,26 @@ def test_length_zero_copy(dtype, copy):
     arr = np.array([], dtype=dtype)
     result = conversion.ensure_datetime64ns(arr, copy=copy)
     assert result.base is (None if copy else arr)
+
+
+class MetaDatetime(type):
+    pass
+
+
+class FakeDatetime(MetaDatetime("NewBase", (datetime,), {})):
+    pass
+
+
+@pytest.mark.parametrize("dt, expected", [
+    pytest.param(Timestamp("2000-01-01"),
+                 Timestamp("2000-01-01", tz=UTC), id="timestamp"),
+    pytest.param(datetime(2000, 1, 1),
+                 datetime(2000, 1, 1, tzinfo=UTC),
+                 id="datetime"),
+    pytest.param(FakeDatetime(2000, 1, 1),
+                 FakeDatetime(2000, 1, 1, tzinfo=UTC),
+                 id="fakedatetime")])
+def test_localize_pydatetime_dt_types(dt, expected):
```

Comment:
Could you test the `CustomBusinessDay` day arithmetic in the original issue as well?

Reply:
Good call on this, in testing the actual arithmetic with a hand-made custom datetime I found a similar problem in `conversion.pyx:convert_datetime_to_tsobject` as well.  It wasn't being hit when testing with the library the user was using as their custom datetimes just happen to implement the `nanosecond` property that regular `datetime` doesn't have.

I'll include that fix in this PR as well if there is no issue with that, since it is the same issue and it was just a quirk of `freezegun` that it didn't show up in that issue.